### PR TITLE
mat: generalise outer and rank update vector parameters

### LINKED ISSUE #167: all: generalise vector functions
Candidate functions and methods at this stage are (only exported functions are listed):

- [x] `stat/statmat.go:func Mahalanobis(x, y *mat.VecDense, chol *mat.Cholesky) float64`
- [x] `mat/cholesky.go:func (c *Cholesky) SolveVec(v, b *VecDense) error`
- [x] `mat/cholesky.go:func (c *Cholesky) SymRankOne(orig *Cholesky, alpha float64, x *VecDense)`ok bool) {
- [x] `mat/symmetric.go:func (s *SymDense) SymRankOne(a Symmetric, alpha float64, x *VecDense)`
- [x] `mat/symmetric.go:func (s *SymDense) RankTwo(a Symmetric, alpha float64, x, y *VecDense)`
- [x] `mat/inner.go:func Inner(x *VecDense, A Matrix, y *VecDense) float64`
- [x] `mat/solve.go:func (v *VecDense) SolveVec(a Matrix, b *VecDense) error`
- [x] `mat/matrix.go:func Dot(a, b *VecDense) float64`
- [x] `mat/vector.go:func (v *VecDense) SliceVec(i, k int) *VecDense`
- [x] `mat/vector.go:func (v *VecDense) CloneVec(a *VecDense)`
- [x] `mat/vector.go:func (v *VecDense) CopyVec(a *VecDense) int`
- [x] `mat/vector.go:func (v *VecDense) ScaleVec(alpha float64, a *VecDense)`
- [x] `mat/vector.go:func (v *VecDense) AddScaledVec(a *VecDense, alpha float64, b *VecDense)`
- [x] `mat/vector.go:func (v *VecDense) AddVec(a, b *VecDense)`
- [x] `mat/vector.go:func (v *VecDense) SubVec(a, b *VecDense)`
- [x] `mat/vector.go:func (v *VecDense) MulElemVec(a, b *VecDense)`
- [x] `mat/vector.go:func (v *VecDense) DivElemVec(a, b *VecDense)`
- [x] `mat/vector.go:func (v *VecDense) MulVec(a Matrix, b *VecDense)`
- [x] `mat/qr.go:func (qr *QR) SolveVec(v *VecDense, trans bool, b *VecDense) error`
- [x] `mat/dense_arithmetic.go:func (m *Dense) RankOne(a Matrix, alpha float64, x, y *VecDense)`
- [x] `mat/dense_arithmetic.go:func (m *Dense) Outer(alpha float64, x, y *VecDense)`
- [x] `mat/dense.go:func (m *Dense) ColView(j int) *VecDense`
- [x] `mat/dense.go:func (m *Dense) RowView(i int) *VecDense`
- [x] `mat/lq.go:func (lq *LQ) SolveVec(v *VecDense, trans bool, b *VecDense) error`
- [x] `mat/lu.go:func (lu *LU) RankOne(orig *LU, alpha float64, x, y *VecDense)`
- [x] `mat/lu.go:func (lu *LU) SolveVec(v *VecDense, trans bool, b *VecDense) error`

Not all are necessarily worth doing or useful (e.g. `ColView`, `RowView`).

### PR DESCRIPTION
**Pull request by kortschak**
Please take a look.

Closes #167.

### CODE REVIEW

**Comment by btracey** [comment]
I got through most of it, but I'm not sure I fully understand our policy on Vectors. For instance, here a lot of things check for the number of columns equaling one. I think instead our policy should be that a `Vector` is always a column vector, and so we can use `Len`. In our current API, `VecDense.T()` returns a `Matrix`, not a `Vector`, and we do not have a `TVec`, like we do in the other types. I think this is a feature.

**Comment by kortschak** [comment]
>In our current API, VecDense.T() returns a Matrix, not a Vector,

This is the nature of the `Matrix` interface, everything that is a matrix must satisfy that.

>and we do not have a TVec

Yes, we do. There is a glitch in the current code that means that you can't `v.T().(Vector)`, but that is an oversight.

**Comment by kortschak** [comment]
The documentation for `VecDense` and `Vector` say they are column vector and vector respectively. One of the issues is that there is only one type of concrete vector and that that is deemed to be column, independent of it's origin, so `m.RowView(i)` returns a vector of some sort, under the model that all vectors are columns, this surprisingly is a transpose of the row. This surprise was mitigated by making the row and column accessors return `Vector` which is merely defined as a column.

The underlying problem is that the Go type system is not strong enough to capture the richness of the operations here without some lenience.

**Comment by btracey** [comment]
>> and we do not have a TVec
> Yes, we do. There is a glitch in the current code that means that you can't v.T().(Vector), but that is an oversight.

You're right, sorry. `VecDense` does not have a `TVec` method (though it probably should).

Let me rephrase my comment, in light of yours. I don't have a problem with saying that a `Vector` is just a vector, and not a column vector. I do think it's a problem to generalize `Outer` and say that the vectors have to be column vectors. There's no need to force that, and it's a bit weird mathematically (why not just insist that y be a row vector?). Instead, I think we should just say `Outer calculates the outer product of the vectors x and y, where both are interpreted as a column vector`. (or, `where x is treated as a column vector and y is treated as a row vector`). This way it doesn't matter if `RowView` returns a row vector or a column vector, you can still use it wherever you use a `Vector`. I'd rather it be, as much as possible, that when the input parameter is a `Vector`, it doesn't matter which orientation (though obviously the orientation does matter when it will be used as `Matrix`).

**Comment by kortschak** [comment]
That sounds reasonable to me; I actually was initially thinking to do that approach, but the documentation was difficult to properly specify so I went the more constrained route that you see here. Note that other PRs have already been merged that have the dimension checks on the input vectors. Because of this, I'd like to merge this as is and have a completely consistent state to more forward from. Do you think that you would be able to do the dimension relaxation PR and add appropriate documentation?

Also see #384 and #385.

**Comment by btracey** [comment]
SGTM. I'll take a look  at the full PR when I get a chance, on travel at the moment.
